### DISCUSSION TITLE
Add monorepo detection for all libraries

### DISCUSSION THREAD

**Discussion by danielcompton** [renovatebot/renovate#26638] [Suggest an Idea]
### Tell us more.

Renovate seems to be able to detect monorepos if the dependencies specify the same SCM URL, e.g. https://github.com/aws/aws-sdk-java-v2.

Currently, users need to manually add presets for every project which is a monorepo so that Renovate can pick it up, e.g. https://github.com/renovatebot/renovate/pull/26631.

What if Renovate automatically grouped together dependencies if they used the same SCM URL without needing config?

There are a few things to consider:

* Security - if anyone can create an update which is treated as part of an existing monorepo, it could be used to smuggle in a dependency update along with other trusted libraries.
  * Mitigation: The malicious dependency already needs to be added to the project, which limits the blast radius of this.
  * Mitigation: I think an additional protection could be to require that in automatic monorepo detection, all the dependencies need the same group ID. E.g. if there was `software.amazon.awssdk/ec2` and `joes.lib/utils` that both referenced https://github.com/aws/aws-sdk-java-v2 as their SCM URL, they wouldn't be grouped together as monorepos. This probably doesn't exist in all supported dependency ecosystems, but I think it would in most.
* Breaking existing setups
  * Mitigation: Make this an opt-in setting
  * Mitigation: Require that dependencies are being upgraded to the same version (not sure if this is a good idea or not, e.g. if there are experimental dependencies + stable dependencies in the same repo, you'd probably still want them all to upgrade together.
  * There could be a new group of presets which disables this for the small number of dependencies where this isn't valid.

**Comment by HonkingGoose** [comment]
I found a old issue that talks about more generic grouping rules for Renovate. Here's a edited quote from the top post of that issue:

> Group all packages:
> 
> 1. that share a source repository.
> 2. that share a GitHub user.
> 3. that share a homepage.
> 4. in a npm organization.

I'll let the Renovate maintainers respond to your ideas. :wink: 

- https://github.com/renovatebot/renovate/issues/11092
